Upgrade operator-sdk to v1.39.1

balintTobik opened the pull request:
<!--
If this is a bug fix, make sure your description includes "Fixes: #xxxx", or
"Closes: #xxxx"

Please provide the following information:
-->
**- Description of the problem which is fixed/What is the use case**
Upgrade operator-sdk to v1.39.1 version.

**- What I did**
Update operator-sdk version in docs/DEVELOPMENT.md
Update versions in Makefile.
Remove codes related to deprecated kube-rbac-proxy.
Add new network-policy related changes.

**- How to verify it**

**- Description for the changelog**
Upgrade operator-sdk to v1.39.1.
Closes: [KATA-3521](https://issues.redhat.com//browse/KATA-3521)
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

littlejawa commented:
/ok-to-test

littlejawa commented:
/retest

littlejawa commented:
As this PR is removing kube-rbac-proxy, it is also addressing #483

bpradipt commented on config/rbac/metrics_auth_role_binding.yaml:
```diff
@@ -1,12 +1,12 @@
-  name: default
-  namespace: system
+  - kind: ServiceAccount
+    name: controller-manager
```
Who creates the `controller-manager` service account?
The manager.yaml doesn't specify `ServiceAccount` explicitly.

balintTobik replied:
Hi @bpradipt! Thanks for the comment!
If I understand correctly this service account was introduced in v1.5.0 sdk version (https://sdk.operatorframework.io/docs/upgrading-sdk-version/v1.5.0/#gov3-add-a-systemcontroller-manager-serviceaccount-to-your-project)
Do you know the "default" was kept by purpose? Or I can change it to new "controller-manager" SA in other rolebinding as well?

bpradipt replied:
@balintTobik I think we just didn't remove it since OSC operator was always installed in a specific namespace.

gkurz reviewed (changes requested):
Hi @balintTobik ! Thanks for working on this.

A new operator SDK release can introduce incompatibilities that call for very specific changes. Every operator SDK release thus comes with a migration guide at https://sdk.operatorframework.io/docs/upgrading-sdk-version/ .

This PR is doing a 3 release jump. Please break this commit down into 3 per-release commits. Each commit should at least provide the link to the corresponding migration guide as in a9d0f2b7118410c5a4a344f9df0fad98dd77fa52 and any extra details that can help. This will greatly improve the reviewing experience and the overall maintainability.

pmores commented:
It would also be easier to review if it were clearer which changes are manual as opposed to generated by the upgrade.

balintTobik committed:
Upgrade to use operator-sdk v1.38.0

Following steps from https://sdk.operatorframework.io/docs/upgrading-sdk-version/v1.38.0/
Skip steps for upgrade 1.37.0, because it doesn't relevant.

Signed-off-by: Balint Tobik <btobik@redhat.com>

balintTobik committed:
Update config/manager/manager.yaml

Update manager container's arguments to make it compatible with the code

Signed-off-by: Balint Tobik <btobik@redhat.com>

balintTobik committed:
Upgrade operator-sdk v1.39.1

Following steps from https://sdk.operatorframework.io/docs/upgrading-sdk-version/v1.39.0/
No migration required upgrading 1.39.0 -> 1.39.1

Signed-off-by: Balint Tobik <btobik@redhat.com>

balintTobik commented:
Thanks for the comments! 
@bpradipt - I've changed it to use "default" as other rolebindings.
@gkurz - I hope I've managed to change it as requested!
@pmores - Sorry, I don't understand what you meant. All these changes are manual changes, none of them are generated.

gkurz commented on Makefile:
```diff
@@ -50,11 +50,11 @@ endif
 # Set the Operator SDK version to use. By default, what is installed on the system is used.
 # This is useful for CI or a project to utilize a specific version of the operator-sdk toolkit.
-OPERATOR_SDK_VERSION ?= v1.36.1
+OPERATOR_SDK_VERSION ?= v1.38.0
```
It is fine to skip v1.37.0 if it doesn't affect us, which seems to be the case from https://sdk.operatorframework.io/docs/upgrading-sdk-version/v1.37.0/, but it should be mentioned explicitly in the commit message for extra clarity.

balintTobik replied:
Thanks! I forget that.

gkurz commented on config/manager/manager.yaml:
```diff
@@ -59,7 +59,8 @@ spec:
           args:
-            - --enable-leader-election
+            - --metrics-bind-address=127.0.0.1:8080
+            - --leader-elect
```
This looks like this is fixing a bug from cbebd561bfa0a3967db447bb051cc451e4447e65 that turned `enable-leader-election` into `leader-elect`. This was a long long time ago so I am a bit surprised this wasn't discovered before. How did you spot it ? Code reading ? Hit some error ? Also, is `metrics-bind-address` needed here ?

Cc @bpradipt as author of cbebd561bfa0a3967db447bb051cc451e4447e65.

balintTobik replied:
When I tested my change, I faced an issue, try to use a non-existing argument. I saw that the it's changed in clusterserviceversion.yaml during bundle creation. It's interesting, because when I build it without these changes (with 1.36.0 operator-sdk), it didn't changed this arguments. 
Maybe, because of removing/adding manifests to clusterserviceversion reload manager manifest as well...

Default, metrics-bind-address is ":8080", so I think it's needed. 
Also this value is used currently: https://github.com/openshift/sandboxed-containers-operator/blob/devel/bundle/manifests/sandboxed-containers-operator.clusterserviceversion.yaml#L373

bpradipt replied:
I wonder how it worked till now with the incorrect option in manager.yaml - https://github.com/openshift/sandboxed-containers-operator/blob/devel/main.go#L97-L98
Only possibility is that the config/manager/manager.yaml never got used.
Anyways, thanks @balintTobik for fixing this.

bpradipt reviewed (approved):
/lgtm
We should test the changes to ensure no breakage. Hope the upstream test will be sufficient.

littlejawa commented:
/hold
As discussed on slack: we want to make build 11 before merging this. 
We can continue the reviewing process, just putting the "hold" label to avoid merging it too soon.

wainersm reviewed (approved):
LGTM

littlejawa commented:
@balintTobik - It seems the kube-rbac-proxy image is still referenced in the bundle CSV file.
You probably need to run `make bundle` and commit the relevant modifications.

We also need to remove it from the manifest file (/config/manifests/bases/sandboxed-containers-operator.clusterserviceversion.yaml)

gkurz commented:
> @balintTobik - It seems the kube-rbac-proxy image is still referenced in the bundle CSV file. You probably need to run `make bundle` and commit the relevant modifications.
> 
> We also need to remove it from the manifest file (/config/manifests/bases/sandboxed-containers-operator.clusterserviceversion.yaml)

Just a reminder that we also need to convert the bundle in midstream.

balintTobik committed:
Remove kube-rbac-proxy container definition

Signed-off-by: Balint Tobik <btobik@redhat.com>

balintTobik committed:
Update files generated by 'make bundle'

Signed-off-by: Balint Tobik <btobik@redhat.com>

gkurz commented on docs/DEVELOPMENT.md:
```diff
@@ -2,11 +2,11 @@
 export ARCH=$(case $(uname -m) in x86_64) echo -n amd64 ;; aarch64) echo -n arm64 ;; *) echo -n $(uname -m) ;; esac)
 export OS=$(uname | awk '{print tolower($0)}')
-export OPERATOR_SDK_DL_URL=https://github.com/operator-framework/operator-sdk/releases/download/v1.36.1
+export OPERATOR_SDK_DL_URL=https://github.com/operator-framework/operator-sdk/releases/download/v1.39.1
```
We don't need to download the SDK manually since 0a1028a134bdb4458df717a6aaedd7ab49cfbe8a but the documentation wasn't updated accordingly.

Please drop this block code entirely in a preparatory patch.

gkurz replied:
We don't need to download the SDK manually since 0a1028a134bdb4458df717a6aaedd7ab49cfbe8a but the documentation wasn't updated accordingly.

Please drop this block code entirely in a preparatory patch (ideally first patch in the PR).

balintTobik committed:
Remove operator-sdk install instuction

No need to install operator-sdk manually

Signed-off-by: Balint Tobik <btobik@redhat.com>

gkurz reviewed (approved):
/lgtm

Thanks @balintTobik !

Let's wait for CI to run and we'll merge this.

littlejawa commented:
/unhold

littlejawa reviewed (approved):
/lgtm

littlejawa commented:
/retest
